tapdb/universe_stats.go

```diff
@@ -401,39 +726,59 @@ func (u *UniverseStats) QuerySyncStats(ctx context.Context,
 		Query: q,
 	}
 
-	readTx := NewUniverseStatsReadTx()
-	err := u.db.ExecTx(ctx, &readTx, func(db UniverseStatsStore) error {
-		// First, we'll map the external query to our SQL specific
-		// struct. We'll need to use the proper null types so the query
-		// works as expected.
-		query := UniverseStatsQuery{
-			AssetName: sqlStr(q.AssetNameFilter),
-			AssetType: func() sql.NullInt16 {
-				if q.AssetTypeFilter == nil {
-					return sql.NullInt16{}
-				}
+	// First, check the cache to see if we already have a cached result for
+	// this query.
+	syncSnapshots := u.syncStatsCache.fetchQuery(q)
+	if syncSnapshots != nil {
+		resp.SyncStats = syncSnapshots
+		return resp, nil
+	}
 
-				return sqlInt16(*q.AssetTypeFilter)
-			}(),
-			SortBy:        sqlStr(sortTypeToOrderBy(q.SortBy)),
-			SortDirection: sqlInt16(q.SortDirection),
-			NumOffset:     int32(q.Offset),
-			NumLimit: func() int32 {
-				if q.Limit == 0 {
-					return universe.MaxPageSize
-				}
+	// Otherwise, we'll grab the main mutex so we can qury the db then
+	// cache the result.
+	u.syncStatsMtx.Lock()
+	defer u.syncStatsMtx.Unlock()
 
-				return int32(q.Limit)
-			}(),
-		}
+	// Check again to see if the value was loaded in while we were waiting.
+	syncSnapshots = u.syncStatsCache.fetchQuery(q)
+	if syncSnapshots != nil {
+		resp.SyncStats = syncSnapshots
+		return resp, nil
+	}
 
-		// In order for the narg clause to work properly, we'll only
-		// apply the asset ID if it's set.
-		var zeroID asset.ID
-		if q.AssetIDFilter != zeroID {
-			query.AssetID = q.AssetIDFilter[:]
-		}
+	// First, we'll map the external query to our SQL specific struct.
+	// We'll need to use the proper null types so the query works as
+	// expected.
+	query := UniverseStatsQuery{
+		AssetName: sqlStr(q.AssetNameFilter),
+		AssetType: func() sql.NullInt16 {
+			if q.AssetTypeFilter == nil {
+				return sql.NullInt16{}
+			}
+
+			return sqlInt16(*q.AssetTypeFilter)
+		}(),
+		SortBy:        sqlStr(sortTypeToOrderBy(q.SortBy)),
+		SortDirection: sqlInt16(q.SortDirection),
+		NumOffset:     int32(q.Offset),
+		NumLimit: func() int32 {
+			if q.Limit == 0 {
+				return int32(math.MaxInt32)
```

Comment:
Is this max because the stats pagination interprets limit differently from the latest pagination PR?

Reply:
Yeah this is kinda legacy. I think we want to enforce another limit here as well. For the stats site it relies on this not being populated, but a future deploy will start to end the limit itself.